From 32c89a7b87523247067ac2c2bc4ae893532e970c Mon Sep 17 00:00:00 2001
From: John Ogness <john.ogness@linutronix.de>
Date: Mon, 23 Jan 2017 21:00:44 +0100
Subject: [PATCH] debian: workaround broken installkernel in jessie

see: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=812266
for details.

This adds the installkernel script from debianutils-4.7 to
the root directory of the kernelsource and adds a calling
hook to the install.sh script of the supported platforms.

Signed-off-by: John Ogness <john.ogness@linutronix.de>
---
 arch/arm/boot/install.sh     |  1 +
 arch/arm64/boot/install.sh   |  1 +
 arch/powerpc/boot/install.sh |  1 +
 arch/x86/boot/install.sh     |  1 +
 elbe_installkernel           | 89 ++++++++++++++++++++++++++++++++++++++++++++
 5 files changed, 93 insertions(+)
 create mode 100644 elbe_installkernel

diff --git a/arch/arm/boot/install.sh b/arch/arm/boot/install.sh
index 2a45092..991f065 100644
--- a/arch/arm/boot/install.sh
+++ b/arch/arm/boot/install.sh
@@ -36,6 +36,7 @@ verify "$3"
 
 # User may have a custom install script
 if [ -x ~/bin/${INSTALLKERNEL} ]; then exec ~/bin/${INSTALLKERNEL} "$@"; fi
+if [ -f ../../elbe_installkernel ]; then exec sh ../../elbe_installkernel "$@"; fi
 if [ -x /sbin/${INSTALLKERNEL} ]; then exec /sbin/${INSTALLKERNEL} "$@"; fi
 
 if [ "$(basename $2)" = "zImage" ]; then
diff --git a/arch/arm64/boot/install.sh b/arch/arm64/boot/install.sh
index d91e1f0..789cd1e 100644
--- a/arch/arm64/boot/install.sh
+++ b/arch/arm64/boot/install.sh
@@ -36,6 +36,7 @@ verify "$3"
 
 # User may have a custom install script
 if [ -x ~/bin/${INSTALLKERNEL} ]; then exec ~/bin/${INSTALLKERNEL} "$@"; fi
+if [ -f ../../elbe_installkernel ]; then exec sh ../../elbe_installkernel "$@"; fi
 if [ -x /sbin/${INSTALLKERNEL} ]; then exec /sbin/${INSTALLKERNEL} "$@"; fi
 
 if [ "$(basename $2)" = "Image.gz" ]; then
diff --git a/arch/powerpc/boot/install.sh b/arch/powerpc/boot/install.sh
index b6a256b..472da56 100644
--- a/arch/powerpc/boot/install.sh
+++ b/arch/powerpc/boot/install.sh
@@ -24,6 +24,7 @@ set -e
 # User may have a custom install script
 
 if [ -x ~/bin/${INSTALLKERNEL} ]; then exec ~/bin/${INSTALLKERNEL} "$@"; fi
+if [ -f ../../elbe_installkernel ]; then exec sh ../../elbe_installkernel "$@"; fi
 if [ -x /sbin/${INSTALLKERNEL} ]; then exec /sbin/${INSTALLKERNEL} "$@"; fi
 
 # Default install
diff --git a/arch/x86/boot/install.sh b/arch/x86/boot/install.sh
index d13ec1c..7c5263b 100644
--- a/arch/x86/boot/install.sh
+++ b/arch/x86/boot/install.sh
@@ -34,6 +34,7 @@ verify "$3"
 # User may have a custom install script
 
 if [ -x ~/bin/${INSTALLKERNEL} ]; then exec ~/bin/${INSTALLKERNEL} "$@"; fi
+if [ -f ../../elbe_installkernel ]; then exec sh ../../elbe_installkernel "$@"; fi
 if [ -x /sbin/${INSTALLKERNEL} ]; then exec /sbin/${INSTALLKERNEL} "$@"; fi
 
 # Default install - same as make zlilo
diff --git a/elbe_installkernel b/elbe_installkernel
new file mode 100644
index 0000000..3aef978
--- /dev/null
+++ b/elbe_installkernel
@@ -0,0 +1,89 @@
+#!/bin/sh
+# 
+# Copyright (C) 1995 - 1998, Ian A. Murdock <imurdock@debian.org>
+# Copyright (C) 1998, 1999, Guy Maor
+# Copyright (C) 2002, Matthew Wilcox
+# Copyright (C) 2002, 2004, 2005, 2007, 2009  Clint Adams
+# Copyright (C) 2009  Manoj Srivasta
+#
+# Install the kernel on a Debian Linux system.
+#
+# This script is called from /usr/src/linux/arch/i386/boot/install.sh.
+# If you install it as /sbin/installkernel, you can do a "make install"
+# from a generic kernel source tree, and the image will be installed to
+# the proper place for Debian GNU/Linux.
+
+set -e
+
+# Parse the command line options.  Of course, powerpc has to be all
+# different, and passes in a fifth argument, just because it is
+# "special". We ignore the fifth argument, and do not flag is as an
+# error, which it would be for any arch apart from powerpc
+if [ $# -eq 3 ] || [ $# -eq 4 ] || [ $# -eq 5 ] ; then
+  img="$2"
+  map="$3"
+  ver="$1"
+  if [ $# -ge 4 ] && [ -n "$4" ] ; then
+      dir="$4"
+  else
+      dir="/boot"
+  fi
+else
+  echo "Usage: installkernel <version> <image> <System.map> <directory>"
+  exit 1
+fi
+
+# Create backups of older versions before installing
+updatever () {
+  if [ -f "$dir/$1-$ver" ] ; then
+    mv "$dir/$1-$ver" "$dir/$1-$ver.old"
+  fi
+
+  cat "$2" > "$dir/$1-$ver"
+
+  # This section is for backwards compatibility only
+  if test -f "$dir/$1" ; then
+    # The presence of "$dir/$1" is unusual in modern intallations, and
+    # the results are mostly unused.  So only recreate them if they
+    # already existed.
+    if test -L "$dir/$1" ; then
+        # If we were using links, continue to use links, updating if
+        # we need to.
+        if [ "$(readlink -f ${dir}/${1})" = "${dir}/${1}-${ver}" ]; then
+            # Yup, we need to change
+            ln -sf "$1-$ver.old" "$dir/$1.old"
+        else
+            mv "$dir/$1" "$dir/$1.old"
+        fi
+        ln -sf "$1-$ver" "$dir/$1"
+    else                        # No links
+        mv "$dir/$1" "$dir/$1.old"
+        cat "$2" > "$dir/$1"
+    fi
+  fi
+}
+
+if [ "$(basename $img)" = "vmlinux" ] ; then
+  img_dest=vmlinux
+else
+  img_dest=vmlinuz
+fi
+updatever $img_dest "$img"
+updatever System.map "$map"
+
+config=$(dirname "$map")
+config="${config}/.config"
+if [ -f "$config" ] ; then
+  updatever config "$config"
+fi
+
+# If installing in the usual directory, run the same scripts that hook
+# into kernel package installation.  Also make sure the PATH includes
+# /usr/sbin and /sbin, just as dpkg would.
+if [ "$dir" = "/boot" ]; then
+  PATH="$PATH:/usr/sbin:/sbin" \
+    run-parts --verbose --exit-on-error --arg="$ver" --arg="$dir/$img_dest-$ver" \
+    /etc/kernel/postinst.d
+fi
+ 
+exit 0
-- 
2.1.4

